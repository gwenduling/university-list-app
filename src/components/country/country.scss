.country-component {
  margin-top: 20px;
}

.result-count {
  font-size: 14px;
}

.show-more {
  width: calc(25% - 10px);

  @media (max-width: 1300px) {
    width: calc(33% - 3px);
  }

  @media (max-width: 500px) {
    width: 100%;
  }
}
