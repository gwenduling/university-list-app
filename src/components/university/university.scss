.university-component {
  padding: 10px;
  color: var(--color-darker);
  background-color: var(--color-light);

  .name {
    margin: 0;
    font-weight: 700;
  }

  .link {
    color: blue;
    font-size: 12px;
  }
}
