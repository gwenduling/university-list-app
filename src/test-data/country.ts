import { CountryModel } from "../models/countries";
export const mockCountry: CountryModel = {
  alpha2Code: "PH",
  altSpellings: ["PH", "Republic of the Philippines", "Rep√∫blika ng Pilipinas"],
  flag: "https://restcountries.eu/data/phl.svg",
  name: "Philippines",
  nativeName: "Pilipinas",
  count: 0,
  total: 0,
};

export const mockCountry2: CountryModel = {
  alpha2Code: "US",
  altSpellings: ["US", "USA", "United States of America"],
  flag: "https://restcountries.eu/data/phl.svg",
  name: "United States of America",
  nativeName: "United States",
  count: 0,
  total: 0,
};
