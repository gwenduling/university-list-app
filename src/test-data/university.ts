import { UniversityModel } from "../models/universities";

export const mockUniversity: UniversityModel = {
  alpha_two_code: "PH",
  country: "Philippines",
  name: "University Test 1",
  domains: ["university.com"],
  web_pages: ["http://university.com"],
  "state-province": null,
};

export const mockUniversity2: UniversityModel = {
  alpha_two_code: "PH",
  country: "Philippines",
  name: "University Test 2",
  domains: ["university2.com"],
  web_pages: ["http://university2.com"],
  "state-province": null,
};
