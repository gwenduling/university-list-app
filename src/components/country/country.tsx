import { Component } from "react";
import { CountryModel } from "../../models/countries";
import { UniversityModel } from "../../models/universities";

import University from "../university/university";
import CountryItem from "../country-item/country-item";

import "./country.scss";

class Country extends Component<Props> {
  renderUniversity(university: UniversityModel) {
    return (
      <University
        university={university}
        key={university.alpha_two_code + "" + university.name}
      />
    );
  }

  renderButton() {
    return (
      <button
        className="show-more"
        onClick={() => this.props.onClick?.(this.props.country.alpha2Code)}
      >
        show more
      </button>
    );
  }

  render() {
    const universities = this.props.universities
      .filter(
        (university: UniversityModel) =>
          university.alpha_two_code === this.props.country.alpha2Code
      )
      .slice(0, this.props.country.count);

    let button;
    if (this.props.country.count < this.props.country.total) {
      button = this.renderButton();
    } else {
      button = "";
    }

    return (
      <div className="country-component">
        <CountryItem country={this.props.country} />

        <div className="result-count">
          Showing {this.props.country.count} of {this.props.country.total}{" "}
          universities
        </div>

        <div className="universities-list">
          {universities.map((university: UniversityModel) => {
            return this.renderUniversity(university);
          })}
        </div>

        {button}
      </div>
    );
  }
}

interface Props {
  country: CountryModel;
  universities: UniversityModel[];
  onClick?: (code: string) => void;
}
export default Country;
