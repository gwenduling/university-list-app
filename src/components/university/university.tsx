import { UniversityModel } from "../../models/universities";

import "./university.scss";

function University(props: Props) {
  return (
    <div className="university-component">
      <p className="name">{props.university.name}</p>
      <a
        className="link"
        href={props.university.web_pages[0]}
        target="_blank"
        rel="noreferrer"
        title={props.university.name}
      >
        {props.university.domains[0]}
      </a>
    </div>
  );
}

interface Props {
  university: UniversityModel;
}
export default University;
