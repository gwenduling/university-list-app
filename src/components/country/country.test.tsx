import Country from "./country";
import University from "../university/university";

import { shallow } from "enzyme";
import { mockCountry } from "../../test-data/country";
import { mockUniversity, mockUniversity2 } from "../../test-data/university";

describe("Country Component", () => {
  it("should match the snapshot with empty universities", () => {
    const wrapper = shallow(
      <Country country={mockCountry} universities={[]} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should render all universities", () => {
    const country = { ...mockCountry, count: 2, total: 2 };
    const wrapper = shallow(
      <Country
        country={country}
        universities={[mockUniversity, mockUniversity2]}
      />
    );

    expect(wrapper.find(University).length).toBe(2);
  });

  it("should only render number universities from count property", () => {
    const country = { ...mockCountry, count: 1, total: 2 };
    const wrapper = shallow(
      <Country
        country={country}
        universities={[mockUniversity, mockUniversity2]}
      />
    );

    expect(wrapper.find(University).length).toBe(1);
  });

  it("should trigger click event", () => {
    const mockCallBack = jest.fn();

    const country = { ...mockCountry, count: 1, total: 2 };
    const wrapper = shallow(
      <Country
        onClick={() => mockCallBack()}
        country={country}
        universities={[mockUniversity, mockUniversity2]}
      />
    );

    wrapper.find("button").simulate("click");
    expect(mockCallBack).toHaveBeenCalledTimes(1);
  });
});
