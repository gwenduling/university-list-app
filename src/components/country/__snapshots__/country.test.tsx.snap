// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Country Component should match the snapshot with empty universities 1`] = `
<div
  className="country-component"
>
  <CountryItem
    country={
      Object {
        "alpha2Code": "PH",
        "altSpellings": Array [
          "PH",
          "Republic of the Philippines",
          "RepÃºblika ng Pilipinas",
        ],
        "count": 0,
        "flag": "https://restcountries.eu/data/phl.svg",
        "name": "Philippines",
        "nativeName": "Pilipinas",
        "total": 0,
      }
    }
  />
  <div
    className="result-count"
  >
    Showing 
    0
     of 
    0
     
    universities
  </div>
  <div
    className="universities-list"
  />
</div>
`;
